### PR TITLE
Core: Add OAuth2 helpers for REST catalog

### PR DESCRIPTION
**Pull request by rdblue**
This adds the OAuth2 classes needed by the REST catalog, from #4830.

This also fixes review feedback from that PR: add serialization tests, refactor into `rest.auth` module.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  }
+
+  public static String tokenResponseToJson(OAuthTokenResponse response) {
+    try {
+      StringWriter writer = new StringWriter();
+      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
+      tokenResponseToJson(response, generator);
+      generator.flush();
+      return writer.toString();
+
```
Comment (danielcweeks) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:172]: nit: extra blank line here

**Review by danielcweeks — APPROVED** [review]
+1 (minor formatting comment)

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 1 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java]: Fixed.

**Comment by rdblue** [comment]
I'm going to merge this before CI finishes because the last update was a whitespace change and shouldn't affect the results.

**Pull request merged**